gapis/api/vulkan/transform_command_splitter.go

```diff
@@ -24,961 +24,1067 @@ import (
 	"github.com/google/gapid/gapis/memory"
 )
 
-// commandSplitter is a transform that will re-write command-buffers and insert replacement
-// commands at the correct locations in the stream for downstream transforms to replace.
-// See: https://www.khronos.org/registry/vulkan/specs/1.1-extensions/html/vkspec.html#renderpass
-// and https://www.khronos.org/registry/vulkan/specs/1.1-extensions/html/vkspec.html#pipelines-graphics
-// to understand how/why we have to split these.
+// To split a renderpass and insert insertion command in it,
+// for every subpass in a renderpass, we create 3 renderpasses
+type subpassSplitters struct {
+	enter        VkRenderPass
+	intermediate VkRenderPass
+	exit         VkRenderPass
+}
+
+type renderPassSplitters []subpassSplitters
+
 type commandSplitter struct {
-	lastRequest      api.SubCmdIdx
-	requestsSubIndex []api.SubCmdIdx
+	requestedCmds   []api.SubCmdIdx
+	cmdsOffset      uint64
+	mutationStarted bool
+	allocations     *allocationTracker
+	stateMutator    transform.StateMutator
 
 	readMemoriesForSubmit []*api.AllocResult
-	readMemoriesForCmd    []*api.AllocResult
-	writeMemoriesForCmd   []*api.AllocResult
-	pool                  VkCommandPool
 
-	thisRenderPass    VkCmdBeginRenderPassXArgsʳ
-	currentRenderPass [][3]VkRenderPass
-	thisSubpass       int
+	splitRenderPasses          map[VkRenderPass]renderPassSplitters
+	rewrittenGraphicsPipelines map[VkPipeline]VkPipeline
+	pendingCommandBuffers      []VkCommandBuffer
 
-	splitRenderPasses      map[VkRenderPass][][3]VkRenderPass
-	fixedGraphicsPipelines map[VkPipeline]VkPipeline
-
-	pendingCommandBuffers []VkCommandBuffer
-	stateMutator          transform.StateMutator
-	allocations           *allocationTracker
+	pool VkCommandPool
 }
 
-func NewCommandSplitter(ctx context.Context) *commandSplitter {
+func NewCommandSplitter(ctx context.Context, cmdsOffset uint64) *commandSplitter {
```

Comment:
Maybe we could call it numInitialCommands or something to make it more obvious what it is doing?

Reply:
It can be another reason that causes an offset. Also using initial commands bind an external logic to internal logic. This way they are more separate.